From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Otto=20Pfl=C3=BCger?= <otto.pflueger@abscue.de>
Date: Sat, 1 Mar 2025 14:11:22 +0100
Subject: [PATCH 046/225] power: supply: sc27xx: ignore saved fractional
 capacity
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Some Android drivers store a fractional capacity in the PMIC registers,
which is then persisted across reboots. To allow using this driver in
dual-boot environments, read only the integer part from the register.

Signed-off-by: Otto Pflüger <otto.pflueger@abscue.de>
---
 drivers/power/supply/sc27xx_fuel_gauge.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/power/supply/sc27xx_fuel_gauge.c b/drivers/power/supply/sc27xx_fuel_gauge.c
index 93125ee0a..e7a2d43ce 100644
--- a/drivers/power/supply/sc27xx_fuel_gauge.c
+++ b/drivers/power/supply/sc27xx_fuel_gauge.c
@@ -60,6 +60,7 @@
 #define SC27XX_FGU_MODE_AREA_MASK	GENMASK(15, 12)
 #define SC27XX_FGU_CAP_AREA_MASK	GENMASK(11, 0)
 #define SC27XX_FGU_MODE_AREA_SHIFT	12
+#define SC27XX_FGU_CAP_INTEGER_MASK	GENMASK(7, 0)
 
 #define SC27XX_FGU_FIRST_POWERTON	GENMASK(3, 0)
 #define SC27XX_FGU_DEFAULT_CAP		GENMASK(11, 0)
@@ -282,7 +283,7 @@ static int sc27xx_fgu_read_last_cap(struct sc27xx_fgu_data *data, int *cap)
 	if (ret)
 		return ret;
 
-	*cap = value & SC27XX_FGU_CAP_AREA_MASK;
+	*cap = value & SC27XX_FGU_CAP_INTEGER_MASK;
 	return 0;
 }
 
